feat: expose Redpanda's listener in the docker network

sago2k8 opened the pull request:
<!-- Type of change
Please label this PR with one of the existing labels, depending on the scope of your change
-->
Add a new `WithListener` function to allow adding new listeners to the advertised listeners so connection within the docker networks is possible

## What does this PR do? 

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, etc.
-->
- Added a method to add new listener to the containers
- Validate the listener
- Iterate over the networks and add the new listeners as alias so it can be easily accessible.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Currently there is no way to access the Kafka Api from another container in the same docker network, at least not an obvious one, so as [Redpanda docs suggests](https://docs.redpanda.com/current/manage/security/listener-configuration/#advertise-a-listener) adding an extra advertise listener allows access to the Kafka endpoint from within the docker network so it can be consume from another test-container. 
## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- 

<!-- Recommended
## How to test this PR

Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->


<!-- Optional
## Follow-ups

Add here any thought that you consider could be identified as an actionable step once this PR is merged.
-->

eddumelendez commented on modules/redpanda/mounts/redpanda.yaml.tpl:
```diff
@@ -10,6 +10,13 @@ redpanda:
 
+    # This listener is required to enable connection from within the Docker
+    # network
+    - address: 0.0.0.0
+      name: docker
+      port: 29092
+      authentication_method: {{ .KafkaAPI.AuthenticationMethod }}
```
In java, [adding multiple listeners is allowed](https://github.com/testcontainers/testcontainers-java/blob/main/modules/redpanda/src/main/resources/testcontainers/redpanda.yaml.ftl#L24-L29) because it can be shared through the docker network and via a proxy such as toxiproxy. I think the same approach can be offered in go. For the test, look at this [one](https://github.com/testcontainers/testcontainers-java/blob/main/modules/redpanda/src/test/java/org/testcontainers/redpanda/RedpandaContainerTest.java#L114-L142) using kcat container.

sago2k8 replied:
@eddumelendez I have update the PR to be more aligned with the implementation you linked, mind to check it out ?, thanks

eddumelendez reviewed (commented):
Thanks for your contribution! I've left a suggestion regarding the implementation and how to test it.

mdelapenya commented on modules/redpanda/options.go:
```diff
@@ -106,3 +119,28 @@ func WithTLS(cert, key []byte) Option {
+// networks
+func WithListener(lis string) Option {
+	host, port, err := net.SplitHostPort(lis)
+
```
```suggestion
```

mdelapenya commented on modules/redpanda/redpanda.go:
```diff
@@ -252,6 +260,29 @@ func renderBootstrapConfig(settings options) ([]byte, error) {
 	return bootstrapConfig.Bytes(), nil
 }
 
+// validateListeners validates that the provided listeners are valid and set network alieases for the provided address.
```
```suggestion
// validateListeners validates that the provided listeners are valid and set network aliases for the provided address.
```

mdelapenya commented on modules/redpanda/redpanda.go:
```diff
@@ -98,10 +100,15 @@ func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomize
 	// the very first time you start a cluster.
 	bootstrapConfigPath := filepath.Join(tmpDir, bootstrapConfigFile)
 	bootstrapConfig, err := renderBootstrapConfig(settings)
+
```
Let's have errors next to their evaluation
```suggestion
```

mdelapenya reviewed (changes requested):
Hi @sago2k8, thanks for submitting this feature. It LGTM, although I've added a few comments regarding format (errors and its evaluation) and comments.

Regarding documentation, I'd like to see this new functional option added to the docs here: `docs/modules/redpanda.md`. Please take a look at the `codeinclude` blocks in order to embed existing code snippets into the docs.

Once this is added, I think we can merge.

Cheers!

sago2k8 commented:
> Hi @sago2k8, thanks for submitting this feature. It LGTM, although I've added a few comments regarding format (errors and its evaluation) and comments (not blockers).
> 
> Regarding documentation, I'd like to see this new functional option added to the docs here: `docs/modules/redpanda.md`. Please take a look at the `codeinclude` blocks in order to embed existing code snippets into the docs.
> 
> Once this is added, I think we can merge.
> 
> Cheers!

Thanks for the review @mdelapenya, I think I addressed all your comments and added the documentation. could you give it another look please :)

mdelapenya commented on modules/redpanda/redpanda_test.go:
```diff
@@ -278,6 +293,117 @@ func TestRedpandaWithTLS(t *testing.T) {
+	ctx := context.Background()
+
+  	// 1. Create network 
+	RPNetwork, err := network.New(ctx, network.WithCheckDuplicate())
```
Nit: let's use lowercase here. Please apply it everywhere
```suggestion
	rpNetwork, err := network.New(ctx, network.WithCheckDuplicate())
```

mdelapenya commented on modules/redpanda/redpanda_test.go:
```diff
@@ -278,6 +293,117 @@ func TestRedpandaWithTLS(t *testing.T) {
+	// 6. Consume message from Redpanda
+	_, stdout, err := kcat.Exec(ctx, []string{"kcat", "-b", "redpanda:29092", "-C", "-t", "msgs", "-c", "1"})
+	require.NoError(t, err)
+	out, err := io.ReadAll(stdout)
```
Nit: separate concerns
```suggestion

	out, err := io.ReadAll(stdout)
```

mdelapenya commented on modules/redpanda/redpanda_test.go:
```diff
@@ -5,19 +5,34 @@ import (
 	"github.com/twmb/franz-go/pkg/sasl/scram"
 )
 
+// TestContainersLogger is a logger that implements the testcontainers
+// logger interface.
+type TestContainersLogger struct {
+	// LogPrefix is a string that is prefixed for every individual log line.
+	LogPrefix string
+}
+
+// Accept prints the log to stdout
+func (lc TestContainersLogger) Accept(l testcontainers.Log) {
+	fmt.Print(lc.LogPrefix + string(l.Content))
+}
+
```
Is this used elsewhere?

sago2k8 replied:
Good catch, this is a leftover I used for debugging, gonna remove it.

mdelapenya commented on modules/redpanda/redpanda.go:
```diff
@@ -252,6 +261,29 @@ func renderBootstrapConfig(settings options) ([]byte, error) {
 
+// validateListeners validates that the provided listeners are valid and set network aliases for the provided address.
+// The container must be attached to at least one network.
+func validateListeners(ctx context.Context, settings options, req testcontainers.GenericContainerRequest) error {
```
Not a blocker:

I think this function it's correct, although for me it's doing two things in one: validate the listeners and define the network aliases.

Given the function name is `validate`, I could expect that the second part for setting the networks aliases is applied elsewhere.

I'm not sure how to improve this: a new name for the function (applyListeners??). PTAL at the comment in https://github.com/testcontainers/testcontainers-go/pull/1994/files#diff-ea8e8772fe279375ed6f3310607a32f3a18780f6ebfe9b3f817699b6afbdda4cR24-R35:

> // 4. Validate listeners and add network aliases to every network... 

If we rename the function, I'd change this too to

> // 4. Apply listeners to network aliases if they are valid... 

Wdyt?

sago2k8 replied:
I agree with the new name

mdelapenya replied:
I like it, thanks for understanding

mdelapenya reviewed (commented):
I think we are almost there, please check my comments 🙏

sago2k8 commented:
Just noticed there was a merge conflict in one of my commits, addressed now :)

mdelapenya reviewed (approved):
This LGTM, thanks for your time!

@eddumelendez if you see anything weird, please let's do it in a follow up PR 🙏

eddumelendez commented on modules/redpanda/redpanda.go:
```diff
@@ -252,6 +260,29 @@ func renderBootstrapConfig(settings options) ([]byte, error) {
+			return fmt.Errorf("invalid port on listener %s:%d (must be between 0 and 65535)", listener.Address, listener.Port)
+		}
+
+		for _, network := range req.Networks {
+			req.NetworkAliases[network] = append(req.NetworkAliases[network], listener.Address)
+		}
```
one issue with this approach that Java implementation still have is that when adding Toxiproxy upfront Kafka container and using `0.0.0.0:<random-toxiproxy-port>` as a listener the registration will fail because will try to add 0.0.0.0 as an alias. I forgot to mention that when suggesting the approach initially.

mdelapenya replied:
@sago2k8 @eddumelendez would you like to collaborate in a follow-up PR for this?

sago2k8 replied:
Hey, thanks for your comments, yes I could work on it, do you happen to have a suggestion of a expected behaviour for that specific case
